go/toolchain/toolchains.bzl

```diff
@@ -150,19 +149,22 @@ def _bootstrap(base):
   bootstrap["declare"] = go_bootstrap_toolchain
   return bootstrap
 
-def _default(base):
+def _default(base, version):
   default = dict(base)
   default["name"] = "default-" + base["name"]
   default.pop("declare")
   default["version_constraints"] = []
+  default["match_version"] = version
   return default
 
 _toolchains = _generate_toolchains()
 _label_prefix = "@io_bazel_rules_go//go/toolchain:"
 
-def register_go_toolchains():
+def go_register_toolchains(go_version=DEFAULT_VERSION):
   # Use the final dictionaries to register all the toolchains
   for toolchain in _toolchains:
+    if "match_version" in toolchain and toolchain["match_version"] != go_version:
+        continue
     native.register_toolchains(_label_prefix + toolchain["name"])
 
 def declare_toolchains():
```

Comment:
`declare_toolchains` doesn't have the same filtering logic as `go_register_toolchains`. Does it make sense to declare toolchains that aren't registered? If not, could this logic be moved into `_generate_toolchains` (to avoid appending to `_toolchains`)?

Reply:
It makes sense to still declare them.
The labels are scoped and unique, so they can't cause collisions, and if they are not registered they can have no impact on the build graph.
Always declaring them makes it easy for projects to just register the subset they want, by not calling the main register function and just hand registering some of them by label.